### ISSUE TITLE
Segfaults when running QANet

### ISSUE THREAD

**Issue by odellus** [ROCmSoftwarePlatform/tensorflow#68]
I'm opening this up bc I was told to at 
https://github.com/ROCmSoftwarePlatform/MIOpen/issues/41

I'm using tensorflow-upstream which doesn't have an issues tab so here I am. Only opened the ticket with MIOpen bc the error message comes from MIOpen. 

I'm trying to run QANet 
https://github.com/NLPLearn/QANet

With tensorflow version rocm-1.7.2 built with bazel 0.10.0 from develop-upstream branch.

Set the system level debug flags. Below is truncated output bc I had trouble piping stdout to a file when running first time. Can redo with tee if necessary.

https://gist.github.com/odellus/87e2a382c5d6967b3b48ea5fbdf566a6

**Comment by parallelo** [comment]
Thanks for providing the logs.  These are useful because they tell us which kernels were the last ones to launch before the GPU's memory access fault.  Here are some kernels we'll be investigating:  
```
...
<<hip-api tid:4.69801 4.69801 hipLaunchKernel '_ZN5Eigen8internal15EigenMetaKernelINS_15TensorEvaluatorIKNS_14TensorAssignOpINS_9TensorMapINS_6TensorIfLi1ELi1ElEELi16ENS_11MakePointerEEEKNS_19TensorCwiseBinaryOpINS0_13scalar_sum_opIffEEKNS9_INSA_IKfSC_EEKNS4_INS5_ISC_Li1ELi1ElEELi16ES7_EESG_EESG_EEEENS_9GpuDeviceEEElEEvT_T0_' gridDim:{11264,1,1} groupDim:{1024,1,1} sharedMem:+0 stream:0.1 @81051045244935
  hip-api tid:4.69801 hipModuleLaunchKernel          ret= 0 (hipSuccess)>> +36912 ns
<<hip-api tid:4.69802 hipModuleLaunchKernel (0x504876700, 24, 1, 1, 32, 1, 1, 0, stream:0.1, 0, 0x7ff32cc74790) @81051045301818
<<hip-api tid:4.69802 4.69802 hipLaunchKernel '_ZN10tensorflow7functor37SwapDimension1And2InTensor3UsingTilesIjLi32ELi14ELi32ELb0EEEvPKT_NS0_9DimensionILi3EEEPS2_' gridDim:{768,1,1} groupDim:{32,1,1} sharedMem:+0 stream:0.1 @81051045312069
  hip-api tid:4.69802 hipModuleLaunchKernel          ret= 0 (hipSuccess)>> +56253 ns
2018-05-26 21:59:56.017379: I tensorflow/core/kernels/conv_grad_filter_ops.cc:959] running auto-tune for Backward-Filter
...
Memory access fault by GPU node-1 on address 0x58b242000. Reason: Page not present or supervisor privilege.
```

We'll try to reproduce this failure with `tensorflow-upstream` using the current public ROCm release (v1.8.0).

**Comment by odellus** [comment]
I've got some triaging to do on my end will open it back up if it becomes clear I should do so. Thank you!

**Comment by parallelo** [comment]
Quick update:  Just kicked off some QANet training.  Will let it run for a while and see what happens.

**Comment by parallelo** [comment]
On our side, the training is still running successfully.  We'll try to let it run during the rest of the long weekend.  

In the meantime, would you be able to confirm whether your system is able to run a few of these simple examples from our quickstart doc?  (this'll help us understand if your system setup is in good shape or not)
https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/rocm_docs/tensorflow-quickstart.md

**Comment by odellus** [comment]
I rebuilt tensorflow-upstream with bazel-0.13.0 because I had been using 0.10.0, thought that might fix it. I also changed the opt flag when running ./configure because I don't have a haswell. Ended up using `-march=x86-64` because it wasn't hearing any `znver1` business. Other than that and version/location of python everything is default config (just pressed enter).

I also wanted to try it out on a more simple model, so I copied this over from some tensorflow tutorial:
https://gist.github.com/odellus/1c32607d2125f653431494466171922e

The new system debug logging output for the simple example is here
https://gist.github.com/odellus/357d25ca39d085a292b6182524cff4a0

Tried to run QANet again after reinstalling, got this:
https://gist.github.com/odellus/e6bde0fce488ad8b8c28f49acf50b0a6

The model example you sent works. Got a test error of 0.8%. Hmmmm. :confused:

**Comment by parallelo** [comment]
Update after the long weekend:  the training is still running without any errors.  Tested on ROCm 1.8.0 with a 16GB MI-25.  

A good next step for you would be to upgrade to the current public ROCm release (1.8.0).  See if that changes the behavior.  If that still doesn't work, I'll see if I can reproduce on a 8GB GPU.  

Try something like this:  
```
sudo apt purge rocm-dkms rock-dkms compute-firmware rocm-dev
sudo apt update
sudo apt install rocm-dkms rock-dkms compute-firmware rocm-dev
sudo reboot
```

**Comment by parallelo** [comment]
Interestingly, your last log showed a _different_ point of failure from your earlier logs.  Here's what I'm referring to:  
```
<<hip-api tid:4.4451 4.4451 hipLaunchKernel 'sp3AsmConvRxSU' gridDim:{18432,1,1} groupDim:{512,1,1} sharedMem:+0 stream:0.1 @170287038569443
Memory access fault by GPU node-1 on address 0x58ac17000. Reason: Page not present or supervisor privilege.
```
Note the `Asm` in the kernel name, which implies hand-written assembly in MIOpen.  This may be important, and we'll keep our eyes on it.  

Some follow-up questions:  
- Are you are running on a Fiji (gfx803) or Vega (gfx900)?  
- Is this last result repeatable?  i.e. do you always see the `sp3AsmConvRxSU` as the final launched kernel?  

Anyway, let's upgrade to ROCm 1.8.0 first and then we can do some more tests.  

Best, 

Jeff

**Comment by dagamayank** [comment]
@parallelo @odellus - are you both using the *same* versions of TensorFlow? Seems like @odellus is on `tensorflow-upstream` but this discussion is on this repo which is NOT the `tensorflow-upstream` repo.

@odellus - can you please confirm which TF version are you seeing this error in?

**Comment by odellus** [comment]
I'm using version 1.8.0-rc1 of develop-upstream branch of tensorflow-upstream on a gfx803. Sorry to crowd your issues here in the stable version, but wyd? 

I should have mentioned that I varied the batch size to get the different behavior. I bumped it back up to the original 32 from the 8 I was using earlier for that last report on QANet's behavior. With batch size of 32 I get repeatable behavior. It is `sp3AsmConvRxSU` that failed again.

Making the batch size 16 gives the kernel `miog_betac_alphaab` as the one that failed.

batch size 8 runs longer and it's `miog_betac_alphaab` that fails again.
batch size 4 - `sp3AsmConvRxSU`

The reason I originally thought it had to do with memory is that batch size changed the behavior of the program pretty drastically. When I lowered the batch size to 1 it ran for quite some while, but I killed it bc the gradient is going to be too noisy with a batch size that low. Increasing the size of batch decreased how long the program would run before segfaulting.

I will upgrade to newest rocm ASAP and work on duplicating issues. And if I vary the batch size in the future I will be sure to say that explicitly sorry again.

**Comment by odellus** [comment]
yeah I've been digging through MIOpen bc I'm trying to add 3D convolutions (so I need to know how 2D convolutions are done in framework first) and I saw all these *.s files and I was like "what are these?" googled it and it's assembly hahaha. First time I've ever actually read through an assembly file.

**Comment by odellus** [comment]
I can't find `compute-firmware` in the sources and I have rocm.list where it's supposed to be in /etc/... 

Reinstalled rocm, pulled and rebuilt MIOpen on rocm 1.8, all tests passing. QANet gives same error in `sp3AsmConvRxSU` for batch size 32 and `miog_betac_alphaab` for batch size 16. 

Should I try to build MIOpen with OpenCL backend maybe you think? Because I seem to be having some trouble with the `conv_3x3_wheel*.so` binaries included in the MIOpen source code.

The other simple example I linked to is also causing the `sp3AsmConvRxSU` kernel to fail.

**Comment by parallelo** [comment]
Sorry about the wrong instructions -- they have been updated above.  `compute-firmware` no longer exists as a separate package in 1.8.0 -- please purge that one.  

The next step will probably be on our side.  We'll have to try on a `gfx803` to see if we can reproduce the failure there.  If we also see the failure, then we'll be looking closely at the `sp3AsmConvRxSU`'s assembly code for `gfx803`.  

Best, 

Jeff

**Comment by fendiwira** [comment]
Hello,

Is there any update for this issue?

I have similiar issue with below mentioned details:

**Issue :**
While executing Tensorflow code on ROCm platform we are facing seg fault

**Error:**
Memory access fault by GPU node-1 on address ..... Reason: Page not present or supervisor privilege.

**Link Code**
https://github.com/pierluigiferrari/ssd_keras

**Environment info:**
    Operating System: Ubuntu 18.04
    ROCm version: 2.0
    TensorFlow 1.12.0
    GPU: gfx803 (Radeon RX580 8 GB)

**Setting**
HSA_ENABLE_SDMA=0
did not avoid the exception

**Temporer solution**
Sometimes lowering batch size will avoid exception

Thanks and Regards

**Comment by parallelo** [comment]
Hi @fendiwira.  Thanks for the issue report -- this issue looks like a new one.  Can you please open a new issue under our current `ROCmSoftwarePlatform/tensorflow-upstream` repo?  (https://github.com/ROCmSoftwarePlatform/tensorflow-upstream)  Our development team is active over there.  (The `ROCmSoftwarePlatform/tensorflow` repo has been deprecated)
